drivers/sensor/qdec_stm32/qdec_stm32.c

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2022, Valerio Setti <vsetti@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT st_stm32_qdec
+
+/** @file
+ * @brief STM32 family Quadrature Decoder (QDEC) driver.
+ */
+
+#include <errno.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/device.h>
+#include <zephyr/init.h>
+#include <soc.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+
+#include <stm32_ll_tim.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(qdec_stm32, CONFIG_SENSOR_LOG_LEVEL);
+
+/* Device constant configuration parameters */
+struct qdec_stm32_dev_cfg {
+	const struct pinctrl_dev_config *pin_config;
+	struct stm32_pclken pclken;
+	TIM_TypeDef *timer_inst;
+	uint16_t prescaler;
+	uint8_t is_input_polarity_inverted;
+	uint8_t input_filtering_level;
+};
+
+/* Device run time data */
+struct qdec_stm32_dev_data {
+	int32_t position;
+};
+
+static int qdec_stm32_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct qdec_stm32_dev_data *dev_data = dev->data;
+	const struct qdec_stm32_dev_cfg *dev_cfg = dev->config;
+
+	dev_data->position = (int32_t) LL_TIM_GetCounter(dev_cfg->timer_inst);
+
+	return 0;
+}
+
+static int qdec_stm32_get(const struct device *dev, enum sensor_channel chan,
+			struct sensor_value *val)
+{
+	struct qdec_stm32_dev_data *const dev_data = dev->data;
+
+	if (chan == SENSOR_CHAN_ROTATION) {
```

Comment:
Unfortunately I'm not sure I can determine how much the physical device rotated in degrees once I get a certain value of the counter. I mean, the number of edges for an entire rotation might change between various mechanical encoders.
Perhaps adding another identifier of raw type would be better?

Reply:
If sensor supports `SENSOR_CHAN_ROTATION` it should return the value in the specified units, ie degrees. A custom/new channel can be introduced to report in "counts" if needed. The driver may support an attribute to set the counts/revolution ratio so that it can support the standard `SENSOR_CHAN_ROTATION`. If other drivers are not doing this, they should really be fixed as it is a bug.